File: 2015/23/23.jl
if basename(pwd()) == "aoc"
    cd("2015/23")
end

mutable struct Registers
    a::BigInt
    b::BigInt
end
Registers() = Registers(0, 0)

function interpret(filename, rs = Registers())
    program = readlines(filename)
    pc = 1
    while checkbounds(Bool, program, pc)
        line = program[pc]
        if startswith(line, "j")
            if startswith(line, "jmp")
                pc += parse(Int, line[4:end])
            else
                cond = startswith(line, "jie") ? iseven : isone
                x = parse(Int, line[7:end])
                pc += cond(getfield(rs, Symbol(line[5]))) ? x : 1
            end
        else
            r = Symbol(line[5])
            if startswith(line, "hlf")
                setfield!(rs, r, getfield(rs, r) ÷ 2)
            elseif startswith(line, "tpl")
                setfield!(rs, r, getfield(rs, r) * 3)
            elseif startswith(line, "inc")
                setfield!(rs, r, getfield(rs, r) + 1)
            end
            pc += 1
        end
    end
    rs
end

@assert interpret("example.txt").a == 2
interpret("input.txt").b
interpret("input.txt", Registers(1, 0)).b

File: 2015/23/23.pl
#!/usr/bin/env swipl
% usage: swipl 23.pl [FILENAME]

:- module(aoc2015day23, []).
:- module(aoc2015day23).
:- initialization(main, main).
:- use_module(library(clpfd)).
:- use_module(library(dcg/basics)).
:- use_module(library(dcg/high_order)).

load(Instructions) --> sequence(instruction, Instructions), blanks, eos.

instruction(hlf(R))    --> "hlf ", register(R), eol.
instruction(tpl(R))    --> "tpl ", register(R), eol.
instruction(inc(R))    --> "inc ", register(R), eol.
instruction(jmp(O))    --> "jmp ", integer(O),  eol.
instruction(jie(R, O)) --> "jie ", register(R), ", ", integer(O), eol.
instruction(jio(R, O)) --> "jio ", register(R), ", ", integer(O), eol.

register(a) --> "a".
register(b) --> "b".

get_register(a, A, r(A, _)).
get_register(b, B, r(_, B)).

set_register(a, A, r(_, B), r(A, B)).
set_register(b, B, r(A, _), r(A, B)).

hlf(N, 1, R0, R) :-
    get_register(N, X0, R0),
    X #= X0 // 2,
    set_register(N, X, R0, R).

tpl(N, 1, R0, R) :-
    get_register(N, X0, R0),
    X #= X0 * 3,
    set_register(N, X, R0, R).

inc(N, 1, R0, R) :-
    get_register(N, X0, R0),
    succ(X0, X),
    set_register(N, X, R0, R).

jmp(O, O, R, R).

jie(N, O0, O, R, R) :-
    get_register(N, X, R),
    (   X mod 2 =:= 0
    ->  O = O0
    ;   O = 1
    ).

jio(N, O0, O, R, R) :-
    (   get_register(N, 1, R)
    ->  O = O0
    ;   O = 1
    ).

interpret(Instrs, I0, R0, R) :-
    (   nth0(I0, Instrs, Instr)
    ->  call(Instr, I1, R0, R1),
        plus(I0, I1, I2),
        interpret(Instrs, I2, R1, R)
    ;   R = R0
    ).

part_one(Instrs, Result) :- interpret(Instrs, 0, r(0, 0), r(_, Result)).
part_two(Instrs, Result) :- interpret(Instrs, 0, r(1, 0), r(_, Result)).

main([File]) :-
    once(phrase_from_file(load(Input), File)),
    part_one(Input, ResultOne),
    writeln(ResultOne),
    part_two(Input, ResultTwo),
    writeln(ResultTwo).

main([]) :- main(['input.txt']).
